### PR TITLE
feat/nft-analytics

### PR DESCRIPTION
**Pull request by danielailie**

### CODE REVIEW

File: src/common/persistence/timescaledb/entities/sum-daily.entity.ts [hunk 1]
```diff
@@ -0,0 +1,32 @@
+import { PrimaryColumn, ViewColumn, ViewEntity } from 'typeorm';
+
+@ViewEntity({
+  expression: `
+    SELECT
+      time_bucket('1 day', timestamp) AS time, series, key,
+      last(value, timestamp) AS last,sum(value) AS sum
```
Comment (gabrielmatei) [inline src/common/persistence/timescaledb/entities/sum-daily.entity.ts:7]: You can remove `last(value, timestamp) AS last` from the query. I don't see it being used in the entity

File: src/common/persistence/timescaledb/entities/sum-weekly.entity.ts [hunk 2]
```diff
@@ -0,0 +1,33 @@
+import { PrimaryColumn, ViewColumn, ViewEntity } from 'typeorm';
+
+@ViewEntity({
+  expression: `
+    SELECT
+      time_bucket('7 day', timestamp) AS time, series, key,
+      last(value, timestamp) AS last,sum(value) AS sum
```
Comment (gabrielmatei) [inline src/common/persistence/timescaledb/entities/sum-weekly.entity.ts:7]: same here

File: src/common/persistence/timescaledb/migrations/1680617570753-AddDailyAndWeeklyViews.ts [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+
+export class AddDailyAndWeeklyViews1680617570753 implements MigrationInterface {
+  name = 'AddDailyAndWeeklyViews1680617570753';
+  transaction = false;
+
+  public async up(queryRunner: QueryRunner): Promise<void> {
+    await queryRunner.query(`CREATE MATERIALIZED VIEW "sum_weekly" WITH (timescaledb.continuous) AS 
+    SELECT
+      time_bucket('7 day', timestamp) AS time, series, key,
+      last(value, timestamp) AS last,sum(value) AS sum
```
Comment (gabrielmatei) [inline src/common/persistence/timescaledb/migrations/1680617570753-AddDailyAndWeeklyViews.ts:11]: when you remove `last` from queries, don't forget to also remove them from migrations

File: src/utils/analytics.utils.ts [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+  const endDate = moment().utc().toDate();
+
+  return [startDate, endDate];
+};
+
+export const convertBinToTimeResolution = (bin: string): TimeResolution => {
+  switch (bin) {
+    case '30s':
+      return TimeResolution.INTERVAL_30_SECONDS;
+    case '30s':
```
Comment (gabrielmatei) [inline src/utils/analytics.utils.ts:32]: remove this line

File: src/common/persistence/timescaledb/analytics-data.setter.service.ts [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+        .insert()
+        .into(XNftsAnalyticsEntity)
+        .values(this.pendingRecords)
+        .orUpdate(['value'], ['timestamp', 'series', 'key']);
+
+      await query.execute();
+
+      this.pendingRecords = [];
+    } catch (error) {
+      console.log(
```
Comment (gabrielmatei) [inline src/common/persistence/timescaledb/analytics-data.setter.service.ts:41]: use logger

Comment (danielailie) [reply src/common/persistence/timescaledb/analytics-data.setter.service.ts]: done

**Review by gabrielmatei — DISMISSED** [review]

File: src/common/persistence/timescaledb/entities/sum-weekly.entity.ts (hunk 2 above)
Comment (danielailie) [reply src/common/persistence/timescaledb/entities/sum-weekly.entity.ts:7]: I'll do this later on, with the next prs

File: src/common/persistence/timescaledb/migrations/1680617570753-AddDailyAndWeeklyViews.ts (hunk 3 above)
Comment (danielailie) [reply src/common/persistence/timescaledb/migrations/1680617570753-AddDailyAndWeeklyViews.ts:11]: I'll do this later on, with the next prs

File: src/utils/analytics.utils.ts (hunk 4 above)
Comment (danielailie) [reply src/utils/analytics.utils.ts]: done

File: src/common/persistence/timescaledb/entities/sum-daily.entity.ts (hunk 1 above)
Comment (danielailie) [reply src/common/persistence/timescaledb/entities/sum-daily.entity.ts:7]: I'll do this later on, with the next prs

Comment (claudiulataretu) [inline src/common/persistence/timescaledb/entities/sum-daily.entity.ts:7]: I don't think you would need the last value for a volume metric. Only sum over a time bucket would be relevant

File: src/common/persistence/timescaledb/typeorm.config.ts [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+export default new DataSource({
+  type: 'postgres',
+  host: process.env.TIMESCALEDB_URL,
+  port: parseInt(process.env.TIMESCALEDB_PORT),
+  username: process.env.TIMESCALEDB_USERNAME,
+  password: process.env.TIMESCALEDB_PASSWORD,
+  database: process.env.TIMESCALEDB_DATABASE,
+  migrationsTransactionMode: 'each',
+  entities: ['dist/common/**/**/entities/*.entity.{js,ts}'],
+  migrations: ['dist/common/persistence/timescaledb/migrations/*.js'],
```
Comment (claudiulataretu) [inline src/common/persistence/timescaledb/typeorm.config.ts:16]: You might need ssl connection here

File: src/modules/analytics/analytics.service.ts [hunk 7]
```diff
@@ -212,28 +159,81 @@ export class AnalyticsService {
...
-
+    this.data = [];
     for (const rawEvent of events) {
       try {
         let parsedEvent = undefined;
-        if (rawEvent.identifier === 'acceptOffer') {
-          parsedEvent = await this.acceptEventParser.handle(rawEvent);
-        } else {
-          parsedEvent = await this.buyEventHandler.handle(rawEvent, 'hash');
+        switch (rawEvent.identifier) {
```
Comment (claudiulataretu) [inline src/modules/analytics/analytics.service.ts:166]: I would not use the event identifier here. Use the first topic from event which represents exactly the event name. Identifier is the endpoint name that triggered the event and you could have different events triggered from the same endpoint

File: src/modules/analytics/trending/trending-collections.service.ts [hunk 8]
```diff
@@ -0,0 +1,239 @@
...
+        key: token,
+        usdPrice: tokenData.priceUsd,
+        decimals: tokenData.decimals,
+      });
+    }
+    return tokensWithPrice.groupBy((t) => t.key);
+  }
+
+  async getFilterAddresses(): Promise<void> {
+    this.filterAddresses = [];
```
Comment (claudiulataretu) [inline src/modules/analytics/trending/trending-collections.service.ts:197]: you don't need this

File: src/modules/analytics/analytics.service.ts (hunk 7 above)
Comment (danielailie) [reply src/modules/analytics/analytics.service.ts:166]: that's right, but I have mapped the enums based on the endpoint name.

File: src/modules/analytics/trending/trending-collections.service.ts (hunk 8 above)
Comment (danielailie) [reply src/modules/analytics/trending/trending-collections.service.ts:197]: done

File: src/common/persistence/timescaledb/timescaledb.module.ts [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+      name: 'timescaledb',
+      useFactory: (apiConfig: ApiConfigService) => ({
+        autoLoadEntities: true,
+        type: 'postgres',
+        host: apiConfig.getTimescaleDbHost(),
+        port: apiConfig.getTimescaleDbPort(),
+        database: apiConfig.getTimescaleDbDatabase(),
+        username: apiConfig.getTimescaleDbUsername(),
+        password: apiConfig.getTimescaleDbPassword(),
+        // ssl: true,
```
[the remark this answers is not in the document]
Comment (danielailie) [reply src/common/persistence/timescaledb/timescaledb.module.ts]: done

File: src/common/persistence/timescaledb/typeorm.config.ts (hunk 6 above)
Comment (danielailie) [reply src/common/persistence/timescaledb/typeorm.config.ts:16]: for the moment seems it is okay like this, I'll update it when needed

**Review by claudiulataretu — DISMISSED** [review]

**Review by gabrielmatei — APPROVED** [review]

**Review by claudiulataretu — APPROVED** [review]

**Pull request merged**